juju/juju
# [JUJU-3648] Controller model table

The following implements the controller model table via the model 
manager domain. The purpose of this set of changes is to ensure 
that we restrict how databases are created. This should then limit 
free-form database creations which could cause problems at scale. 
The solution is to add the model to the controller model table and 
when requesting to access that model, that the uuid exists in that table.

Destroying a model will also remove the model from the controller model
list table. If the model fails to be removed from the db, but is removed
from state, then it will still be cleaned up. The model uuid will be
tainted and can not be used again. This should be fine as model uuid are
randomised.

Additionally, there is a namespace cache, so that we don't have to
verify the model namespace exists for a given duration (roughly 1
minute). As the model for getting a tracked db based on a model is
now late binding, we'll need to potentially request the db from the
apiserver level and this is to prevent a thundering herd problem on
start up.

The code is actually quite logical, if the namespace is the controller
namespace, then we perform no checks, everything else is fair game.

The model manager domain was also partially implemented to allow for
this to happen. The feedback here was the result of the sprint discussion
about where the state should live.

We should do something about UUIDs in juju there are 6 packages for uuid
imported. We should pick one and use that. I'm currently using utils/uuid
but would rather use an external one, or anything not within the utils
package.

## Checklist



## QA steps

Adding a model should show the new model.

```sh
$ juju bootstrap lxd test --build-agent
$ juju add-model default
$ make repl-install repl
$ dqlite -s $(sudo cat /var/lib/juju/dqlite/info.yaml | grep "Address: " | cut -d" " -f2) controller
$ SELECT * FROM model_list;
```

Destroying a model should remove the model.

```sh
$ juju destroy-model default --no-prompt
$ make repl-install repl
$ dqlite -s $(sudo cat /var/lib/juju/dqlite/info.yaml | grep "Address: " | cut -d" " -f2) controller
$ SELECT * FROM model_list;
```
